From 600b5bb856eced5496779936e213c9f364d66ed3 Mon Sep 17 00:00:00 2001
From: Noam Camus <ncamus@habana.ai>
Date: Tue, 16 Apr 2019 16:47:04 +0300
Subject: [PATCH] [SW-0] uart: update BAUDDIV

Update bauddiv register and do not count on BTL value
This is true since we may change PLL which UART depends on (PCI)

Change-Id: I035c46e2a4dd8f8c23d8c07e7cd898b33818926e
---

diff --git a/common/preboot.c b/common/preboot.c
index 08be6c5..d3efa8c 100644
--- a/common/preboot.c
+++ b/common/preboot.c
@@ -162,6 +162,7 @@
 	preboot_delay();
 	preboot_sync_btl();
 	hl_pll_init();
+	hl_uart_init();
 	preboot_copy_spl();
 
 	spl_start();
diff --git a/common/preboot.h b/common/preboot.h
index 701661f..b7515cf 100644
--- a/common/preboot.h
+++ b/common/preboot.h
@@ -45,5 +45,6 @@
 }
 
 u64 get_cpu_boot_status_reg(void);
+void hl_uart_init(void);
 
 #endif /* COMMON_PREBOOT_H_ */
diff --git a/common/uart.c b/common/uart.c
index 88f2b9b..92d729c 100644
--- a/common/uart.c
+++ b/common/uart.c
@@ -13,6 +13,11 @@
 #define UART_SR_TXFULL			(1 << 0) /* TX Buffer Full */
 #define STR_UINT_SZ			8
 
+void preboot_uart_init(unsigned int val)
+{
+	regmap_writel(val, mmPSOC_UART_0_BAUDDIV);
+}
+
 void preboot_puts_raw(const char *str)
 {
 	int i;
diff --git a/common/uart.h b/common/uart.h
index 4935d05..56702b9 100644
--- a/common/uart.h
+++ b/common/uart.h
@@ -10,6 +10,7 @@
 #ifndef _HL_UART_H_
 #define _HL_UART_H_
 
+void preboot_uart_init(unsigned int val);
 void preboot_puts_raw(const char *str);
 void preboot_puts(const char *str);
 int preboot_uint2hexstring(char *str, unsigned int num, bool pad);
diff --git a/gaudi/gaudi.c b/gaudi/gaudi.c
index 5224dce..1eb4088 100644
--- a/gaudi/gaudi.c
+++ b/gaudi/gaudi.c
@@ -10,6 +10,7 @@
 
 #include "preboot.h"
 #include "pll.h"
+#include "uart.h"
 
 void preboot_pcie_wait_for_link(void)
 {
@@ -54,3 +55,7 @@
 	return pll_num;
 }
 
+void hl_uart_init(void)
+{
+	preboot_uart_init(UART_BAUDDIV);
+}
diff --git a/gaudi/gaudi.h b/gaudi/gaudi.h
index 4ae076f..e2782d6 100644
--- a/gaudi/gaudi.h
+++ b/gaudi/gaudi.h
@@ -12,6 +12,9 @@
 #define mmPSOC_PCI_PLL_BASE		0xC72000
 #define CPU_PLL_BASE			mmPSOC_CPU_PLL_BASE
 
+#define PCI_DIV2_CLOCK			(830000000 / 5)
+#define UART_BAUDDIV			(PCI_DIV2_CLOCK / 115200)
+
 void preboot_pcie_wait_for_link(void);
 static inline void preboot_delay(void) {}
 static inline void preboot_putc_virt_uart(const char c) {}
diff --git a/goya/goya.c b/goya/goya.c
index c0303cc..afa1dcb 100644
--- a/goya/goya.c
+++ b/goya/goya.c
@@ -188,3 +188,6 @@
 
 	return pll_num;
 }
+
+void hl_uart_init(void) {};
+
